butte/src/vector.rs

```diff
@@ -110,47 +118,67 @@ mod le_safe_slice_impls {
 #[cfg(target_endian = "little")]
 pub use self::le_safe_slice_impls::*;
 
-pub fn follow_cast_ref<'a, T: Sized + 'a>(buf: &'a [u8], loc: usize) -> &'a T {
+pub fn follow_cast_ref<'a, T: Sized + 'a>(buf: &'a [u8], loc: usize) -> Result<&'a T, Error> {
     let sz = size_of::<T>();
-    let buf = &buf[loc..loc + sz];
+    let buf = buf.get(loc..loc + sz).ok_or(Error::OutOfBounds)?;
     let ptr = buf.as_ptr() as *const T;
-    unsafe { &*ptr }
+    // magnet: is this safe for any T: Sized? I believe we'd need an extra bound with the required invariants.
```

Comment:
I don't think this is safe for `Sized` either. If `T` is `bool` then this could result in UB if the bytes don't conform to Rust's book representation.

Reply:
@nagisa should maybe take a look at this PR.

Reply:
I added the comment because I wanted to keep track of this.  Unfortunately that behavior is already there and apparently unsafe, to be fixed later.